File: frontend/src/api/marketData.ts
import { apiClient } from "./client";

export interface StockQuote {
  symbol: string;
  price: number | null;
  previous_close: number | null;
  open: number | null;
  day_high: number | null;
  day_low: number | null;
  volume: number | null;
  change: number | null;
  change_percent: number | null;
  currency: string | null;
  market_state: string | null;
  updated_at: string;
}

export interface PricePoint {
  timestamp: string;
  open: number | null;
  high: number | null;
  low: number | null;
  close: number | null;
  volume: number | null;
}

export interface HistoricalPrices {
  symbol: string;
  interval: string;
  points: PricePoint[];
}

export async function getQuote(
  symbol: string,
): Promise<StockQuote> {
  const response =
    await apiClient.get<StockQuote>(
      `/market/quote/${symbol}`,
    );

  return response.data;
}

export async function getHistory(
  symbol: string,
  period = "1mo",
  interval = "1d",
): Promise<HistoricalPrices> {
  const response =
    await apiClient.get<HistoricalPrices>(
      `/market/history/${symbol}`,
      {
        params: {
          period,
          interval,
        },
      },
    );

  return response.data;
}

File: backend/app/schemas/market_data.py
from datetime import datetime

from pydantic import BaseModel


class StockQuote(BaseModel):
    symbol: str
    price: float | None = None
    previous_close: float | None = None
    open: float | None = None
    day_high: float | None = None
    day_low: float | None = None
    volume: int | None = None

    change: float | None = None
    change_percent: float | None = None

    currency: str | None = None

    market_state: str | None = None

    updated_at: datetime


class PricePoint(BaseModel):
    timestamp: datetime
    open: float | None = None
    high: float | None = None
    low: float | None = None
    close: float | None = None
    volume: int | None = None


class HistoricalPrices(BaseModel):
    symbol: str
    interval: str
    points: list[PricePoint]
